Mathlib/Combinatorics/SimpleGraph/Extremal/TuranDensity.lean

```diff
@@ -0,0 +1,127 @@
+  intro G _ h
+  rw [mul_comm, ← mul_div_assoc, le_div_iff₀' (mod_cast Nat.choose_pos hn), Nat.cast_choose_two,
+    Nat.cast_choose_two, Nat.cast_add_one, add_sub_cancel_right (n : ℝ) 1,
+    mul_comm _ (n-1 : ℝ), ← mul_div (n-1 : ℝ), mul_comm _ (n/2 : ℝ), mul_assoc,
+    ← mul_div (n+1 : ℝ), mul_comm _ (n/2 : ℝ), mul_assoc, mul_le_mul_left (by positivity),
+    ← Nat.cast_pred (by positivity), ←Nat.cast_mul, ←Nat.cast_add_one, ←Nat.cast_mul, Nat.cast_le]
+  -- double counting `(v, e) ↦ v ∉ e`
+  let s := (univ ×ˢ G.edgeFinset).filter fun (v, e) ↦ v ∉ e
+  trans #s
+  -- counting `e`
+  · conv_rhs =>
+      rw [card_filter, sum_product_right]
+      enter [2, x]
+      rw [← card_filter]
+    simp_rw [Sym2.mem_toFinset, filter_not, filter_mem_eq_inter, ← sum_attach G.edgeFinset,
+      univ_inter, ← compl_eq_univ_sdiff, card_compl, Fintype.card_fin, card_toFinset_mem_edgeFinset,
+      sum_const, card_attach, smul_eq_mul, Nat.succ_sub_succ_eq_sub, mul_comm]
+    rfl
+  -- counting `v`
+  · conv_rhs =>
+      rw [← Fintype.card_fin (n+1), ← card_univ, ← smul_eq_mul, ← sum_const]
+    conv_lhs =>
+      rw [card_filter, sum_product]
+      enter [2, x]
+      rw [← card_filter]
+    apply sum_le_sum
+    intro i _
+    simpa [edgeFinset_deleteIncidenceSet_eq_filter]
+      using card_edgeFinset_deleteIncidenceSet_le_extremalNumber h i
+
+/-- The limit `extremalNumber (Fin n) H / n.choose 2` as `n` approaches `∞` exists. -/
+lemma exists_tendsto_extremalNumber_div_choose_two (H : SimpleGraph W) :
+    ∃ x, Tendsto (fun (n : ℕ) ↦ (extremalNumber n H / n.choose 2 : ℝ)) atTop (𝓝 x) := by
+  let f := fun (n : ℕ) ↦ (extremalNumber n H / n.choose 2 : ℝ)
+  suffices h : ∃ x, Tendsto (fun (n : ℕ) ↦ f (n + 2)) atTop (𝓝 x) by
+    simpa [tendsto_add_atTop_iff_nat 2] using h
+  use ⨅ n, f (n + 2)
+  apply tendsto_atTop_ciInf
+  · apply antitone_nat_of_succ_le
+    intro n
+    apply extremalNumber_div_choose_two_succ_le
+    exact le_add_of_nonneg_left n.zero_le
+  · use 0
+    intro n ⟨_, hn⟩
+    rw [← hn]
+    positivity
+
+/-- The **Turán density** of a simple graph `H` is the limit of `extremalNumber n H / n.choose 2`
+as `n` approaches `∞`.
+
+See `SimpleGraph.tendsto_turanDensity` for proof of existence. -/
+noncomputable def turanDensity (H : SimpleGraph W) :=
+  limUnder atTop fun (n : ℕ) ↦ (extremalNumber n H / n.choose 2 : ℝ)
+
+/-- The **Turán density** of a simple graph `H` is well-defined. -/
+theorem tendsto_turanDensity (H : SimpleGraph W) :
+    Tendsto (fun (n : ℕ) ↦ (extremalNumber n H / n.choose 2 : ℝ)) atTop (𝓝 (turanDensity H)) := by
+  have ⟨_, h⟩ := exists_tendsto_extremalNumber_div_choose_two H
+  rwa [← Tendsto.limUnder_eq h] at h
+
+/-- `extremalNumber n H` is asymptotically equivalent to `turanDensity H * n.choose 2` as `n`
+approaches `∞`. -/
+theorem isEquivalent_extremalNumber (h : turanDensity H ≠ 0) :
+    (fun (n : ℕ) ↦ (extremalNumber n H : ℝ))
+      ~[atTop] (fun (n : ℕ) ↦ (turanDensity H * n.choose 2 : ℝ)) := by
```

Comment:
Is there really no way to fit this in one line? I'd guess you don't need the type ascription on `n` in the second case, and maybe not even in the first case...?